crates/red_knot/src/semantic/types.rs

```diff
@@ -764,16 +764,49 @@ pub(crate) struct UnionType {
 
 impl UnionType {
     fn display(&self, f: &mut std::fmt::Formatter<'_>, store: &TypeStore) -> std::fmt::Result {
-        f.write_str("(")?;
+        let (int_literals, other_types): (Vec<Type>, Vec<Type>) = self
+            .elements
+            .iter()
+            .copied()
+            .partition(|ty| matches!(ty, Type::IntLiteral(_)));
+        let num_types = usize::from(!int_literals.is_empty()) + other_types.len();
+        if num_types > 1 {
+            f.write_str("(")?;
+        }
         let mut first = true;
-        for ty in &self.elements {
+        if !int_literals.is_empty() {
+            f.write_str("Literal[")?;
+            let mut nums: Vec<i64> = int_literals
+                .into_iter()
+                .filter_map(|ty| {
+                    if let Type::IntLiteral(n) = ty {
+                        Some(n)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+            nums.sort_unstable();
+            for num in nums {
+                if !first {
+                    f.write_str(", ")?;
+                }
+                write!(f, "{num}")?;
+                first = false;
+            }
```

Comment:
Could get rid of the need for the `first` variable here:

```suggestion
            for (i, num) in nums.iter().enumerate() {
                if i != 0 {
                    f.write_str(", ")?;
                }
                write!(f, "{num}")?;
            }
```

Reply:
We also need the `first` variable below, though, to know whether we added any literal type at all. Once you account for that, I don't think it ends up cleaner without the `first` variable.

Reply:
Ah didn't spot that! Ignore this then :)